samples/tab-calendar/ts/src/components/Tab.tsx

```diff
@@ -0,0 +1,380 @@
+    this.callGraphFromClient = this.callGraphFromClient.bind(this);
+    this.showConsentDialog = this.showConsentDialog.bind(this);
+  }
+
+  //React lifecycle method that gets called once a component has finished mounting
+  //Learn more: https://reactjs.org/docs/react-component.html#componentdidmount
+  componentDidMount() {
+    // Initialize the Microsoft Teams SDK
+    microsoftTeams.app.initialize().then(() => {
+      // Get the user context from Teams and set it in the state
+      microsoftTeams.app
+        .getContext()
+        .then((context: microsoftTeams.app.Context) => {
+          this.setState({ context: context });
+        });
+
+      microsoftTeams.authentication
+        .getAuthToken()
+        .then((result: string) => {
+          this.setState({ ssoToken: result });
+          this.exchangeClientTokenForServerToken(result);
+        })
+        .catch((error: string) => {
+          console.error("SSO failed: ", error);
+        });
+    });
+  }
+
+  //Exchange the SSO access token for a Graph access token
+  //Learn more: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow
+  exchangeClientTokenForServerToken = async (token: string) => {
+    let serverURL = `${process.env.REACT_APP_BASE_URL}/getGraphAccessToken?ssoToken=${token}&upn=${this.state.context?.user?.userPrincipalName}`;
+    let response = await fetch(serverURL).catch(this.unhandledFetchError); //This calls getGraphAccessToken route in /api-server/app.js
+    if (response) {
+      let data = await response.json().catch(this.unhandledFetchError);
+      if (!response.ok && data.error === "consent_required") {
+        //A consent_required error means it's the first time a user is logging into to the app, so they must consent to sharing their Graph data with the app.
+        //They may also see this error if MFA is required.
+        this.setState({ consentRequired: true }); //This displays the consent required message.
+        this.showConsentDialog(); //Proceed to show the consent dialogue.
+      } else if (!response.ok) {
+        //Unknown error
+        console.error(data);
+        this.setState({ error: true });
+      } else {
+        this.setState({ allItems: data.value });
+      }
+    }
+  };
+
+  //Show a popup dialogue prompting the user to consent to the required API permissions. This opens ConsentPopup.js.
+  //Learn more: https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/authentication/auth-tab-aad#initiate-authentication-flow
+  showConsentDialog() {
+    microsoftTeams.authentication
+      .authenticate({
+        url: window.location.origin + "/auth-start",
+        width: 600,
+        height: 535,
+      })
+      .then((result: string) => {
+        this.consentSuccess(result ?? "");
+      })
+      .catch((reason: string) => {
+        this.consentFailure(reason ?? "");
+      });
+  }
+
+  //Callback function for a successful authorization
```

Comment:
// Callback function for a successful authorization.

Reply:
Fixed